### PR TITLE
Update sygus v1 parser to use ParseOp utility

### PR DESCRIPTION
**Pull request by ajreynol**
This is in preparation for the parser migration.  

Currently, the migration of the sygus v1 parser is not possible due to the way expressions are generated for operators. This PR updates this parser to use the ParseOp utility, which allows for lazy operator/expression creation and will make things more flexible for the parser migration.

It also updates the applyParseOp/getFunctionKind functions in the smt2 parser to more robust (previously they were only hard coded to work for the special cases we required).

### CODE REVIEW

File: src/parser/smt2/smt2.cpp [hunk 1]
```diff
@@ -1160,11 +1178,10 @@ Type Smt2::processSygusNestedGTerm( int sub_dt_index, std::string& sub_dname, st
...
         }
       }
-    }else{
+    }
+    else
+    {
       std::vector< Expr > children;
-      if( sop.getKind() != kind::BUILTIN ){
-        children.push_back( sop );
-      }
```
Comment (HanielB) [inline src/parser/smt2/smt2.cpp]: Why is it OK to remove this?

Comment (HanielB) [reply src/parser/smt2/smt2.cpp]: Please ignore, I thought I had deleted this

Comment (ajreynol) [reply src/parser/smt2/smt2.cpp]: It was deleted, but is now covered by `applyParseOp` below.

Comment (HanielB) [reply src/parser/smt2/smt2.cpp]: No, I mean I thought I had deleted the comment. :)

Comment (ajreynol) [reply src/parser/smt2/smt2.cpp]: Hahah ok :)

**Comment by 4tXJ7f** [comment]
The following patch should fix the Swig issue:
```
diff --git a/src/bindings/java/CMakeLists.txt b/src/bindings/java/CMakeLists.txt
index 3a88a30a2..3cd6a7e51 100644
--- a/src/bindings/java/CMakeLists.txt
+++ b/src/bindings/java/CMakeLists.txt
@@ -199,7 +199,6 @@ set(gen_java_files
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__unordered_mapT_CVC4__Expr_CVC4__ProofLetCount_CVC4__ExprHashFunction_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__vectorT_CVC4__DatatypeConstructorArg_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__vectorT_CVC4__DatatypeConstructor_t.java
-  ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__vectorT_CVC4__ParseOp_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__vectorT_CVC4__SygusGTerm_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__vectorT_std__pairT_CVC4__Expr_CVC4__Expr_t_t.java
   ${CMAKE_CURRENT_BINARY_DIR}/SWIGTYPE_p_std__vectorT_std__vectorT_std__string_t_t.java
diff --git a/src/expr/datatype.i b/src/expr/datatype.i
index 6bd5eb82c..83e21793c 100644
--- a/src/expr/datatype.i
+++ b/src/expr/datatype.i
@@ -9,6 +9,8 @@
 #endif /* SWIGJAVA */
 %}
 
+%include "expr/kind.i"
+
 %extend std::vector< CVC4::Datatype > {
   /* These member functions have slightly different signatures in
    * different swig language packages.  The underlying issue is that
diff --git a/src/parser/cvc4parser.i b/src/parser/cvc4parser.i
index 2ad3bf01d..accb4826c 100644
--- a/src/parser/cvc4parser.i
+++ b/src/parser/cvc4parser.i
@@ -9,7 +9,8 @@ namespace CVC4 {}
 using namespace CVC4;
 %}
 
-%include "parser/parser_exception.i"
 %include "parser/input.i"
+%include "parser/parse_op.i"
 %include "parser/parser.i"
 %include "parser/parser_builder.i"
+%include "parser/parser_exception.i"
diff --git a/src/parser/parse_op.i b/src/parser/parse_op.i
new file mode 100644
index 000000000..37c3bd30f
--- /dev/null
+++ b/src/parser/parse_op.i
@@ -0,0 +1,5 @@
+%{
+#include "parser/parse_op.h"
+%}
+
+%include "parser/parse_op.h"
```

**Pull request merged**